test/unit/evault/modules/Vault/decimals.t.sol

```diff
@@ -0,0 +1,216 @@
+        assertEq(eTST3.convertToShares(0.800861e6), 0.8e6); //js value = 0.800860e6
+
+        // Try to pay off full amount:
+        startHoax(user3);
+        eTST3.repay(0.302511e6, user3);
+        assertEq(eTST3.debtOf(user3), 0);
+        assertEq(eTST3.debtOfExact(user3), 0);
+
+        // Check if any more interest is accrued after mined block:
+        skip(1);
+        assertEq(eTST3.debtOf(user3), 0);
+        assertEq(eTST3.debtOfExact(user3), 0);
+        assertEq(eTST3.totalBorrows(), 0);
+        assertEq(eTST3.totalBorrowsExact(), 0);
+    }
+
+    //decimals() on e vaults should return same value as underlying
+    function test_decimals_sameValueAsUnderlying() public {
+        assertEq(assetTST3.decimals(), 6);
+        assertEq(eTST3.decimals(), 6);
+    }
+
+    function test_decimals_zeroDecimals() public {
+        // TST4 has 0 decimals
+        assetTST4.mint(user1, 100);
+        startHoax(user1);
+        assetTST4.approve(address(eTST4), type(uint256).max);
+        evc.enableCollateral(user1, address(eTST4));
+        eTST4.deposit(50, user1);
+
+        assertEq(assetTST4.decimals(), 0);
+        assertEq(eTST4.decimals(), 0);
+    }
+
+    //decimals() on d tokens should always return underlying decimals
+    function test_decimals_dToken() public {
+        assertEq(assetTST3.decimals(), 6);
+        assertEq(DToken(eTST3.dToken()).decimals(), 6);
+
+        assetTST4.mint(user1, 100);
+        startHoax(user1);
+        assetTST4.approve(address(eTST4), type(uint256).max);
+        eTST4.deposit(50, user1);
+
+        startHoax(user3);
+        evc.enableController(user3, address(eTST4));
+        eTST4.borrow(1, user3);
+
+        assertEq(assetTST4.decimals(), 0);
+        assertEq(DToken(eTST4.dToken()).decimals(), 0);
+    }
+
+    //no dust left over after max uint redeem
+    function test_redeem_noDustLeft() public {
+        startHoax(user1);
+        eTST3.deposit(1e6, user1);
+        eTST3.withdraw(0.2e6, user1, user1);
+        assertEq(eTST3.totalSupply(), 0.8e6);
+
+        startHoax(user3);
+        evc.enableController(user3, address(eTST3));
+        eTST3.borrow(0.3e6, user3);
+
+        startHoax(address(this));
+        eTST3.setInterestRateModel(address(new IRMTestFixed()));
+        assetTST3.mint(user3, 0.1e6);
+
+        skip(2628000); // 1 month in seconds
+
+        startHoax(user3);
+        eTST3.repay(0.302511e6, user3);
+
+        startHoax(user1);
+        eTST3.redeem(type(uint256).max, user1, user1);
+        assertEq(eTST3.balanceOf(user1), 0);
+    }
+
+    //total supply of underlying
+    function test_totalSupply() public {
+        startHoax(user1);
+        eTST3.deposit(1.5e6, user1);
+
+        assertEq(eTST3.totalSupply(), 1.5e6);
+        assertEq(eTST3.totalAssets(), 1.5e6);
+    }
+
+    function debtExact(uint256 value) internal pure returns (uint256) {
+        return value * (2 ** 31) / (10 ** 9);
```

Comment:
this was a bit dirty in the old repo. The exact debt is no longer scaled by 1e9 (10**9), but only by 2**31 (which can be written as 2 << INTERNAL_DEBT_PRECISION_SHIFT). So the division by 1e9 is not needed here. On the other hand, when `debtExact` is called, amounts are passed in 1e15. TST3 has six decimals, so this 1e15 doesn't make sense, but it did work with the /1e9 division here. No idea how that happened in the old repo. In short - don't divide by 1e9 here, and call with e6 scale, not e15